noodles-util/src/variant/reader/builder.rs

```diff
@@ -0,0 +1,161 @@
+    ///
+    /// ```
+    /// use noodles_util::variant::{self, Compression};
+    /// let builder = variant::reader::Builder::default().set_compression(Compression::Bgzf);
+    /// ```
+    pub fn set_compression(mut self, compression: Compression) -> Self {
+        self.compression = Some(compression);
+        self
+    }
+
+    /// Sets the format of the input.
+    ///
+    /// By default, the format is autodetected on build. This can be used to override it.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use noodles_util::variant::{self, Format};
+    /// let builder = variant::reader::Builder::default().set_format(Format::Vcf);
+    /// ```
+    pub fn set_format(mut self, format: Format) -> Self {
+        self.format = Some(format);
+        self
+    }
+
+    /// Builds a variant reader from a path.
+    ///
+    /// By default, the format and compression will be autodetected. This can be overridden by
+    /// using [`Self::set_format`] and [`Self::set_compression`].
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// # use std::io;
+    /// use noodles_util::variant;
+    /// let reader = variant::reader::Builder::default().build_from_path("sample.vcf")?;
+    /// # Ok::<_, io::Error>(())
+    /// ```
+    pub fn build_from_path<P>(self, path: P) -> io::Result<Reader<Box<dyn BufRead>>>
+    where
+        P: AsRef<Path>,
+    {
+        let file = File::open(path)?;
+        self.build_from_reader(file)
+    }
+
+    /// Builds a variant reader from a reader.
+    ///
+    /// By default, the format and compression will be autodetected. This can be overridden by
+    /// using [`Self::set_format`] and [`Self::set_compression`].
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use std::io;
+    /// use noodles_util::variant;
+    /// let reader = variant::reader::Builder::default().build_from_reader(io::empty())?;
+    /// # Ok::<_, io::Error>(())
+    /// ```
+    pub fn build_from_reader<R>(self, reader: R) -> io::Result<Reader<Box<dyn BufRead>>>
+    where
+        R: Read + 'static,
+    {
+        let mut reader: Box<dyn BufRead> = Box::new(BufReader::new(reader));
+
+        let compression = self
+            .compression
+            .map(Ok)
+            .unwrap_or_else(|| detect_compression(&mut reader))?;
+
+        let format = self
+            .format
+            .map(Ok)
+            .unwrap_or_else(|| detect_format(&mut reader, compression))?;
+
+        let inner: Box<dyn VariantReader<_>> = match (format, compression) {
+            (Format::Vcf, Compression::Uncompressed) => {
+                let inner: Box<dyn BufRead> = Box::new(BufReader::new(reader));
+                Box::new(vcf::Reader::new(inner))
+            }
+            (Format::Vcf, Compression::Bgzf) => {
+                let inner: Box<dyn BufRead> = Box::new(bgzf::Reader::new(reader));
+                Box::new(vcf::Reader::new(inner))
+            }
+            (Format::Bcf, Compression::Uncompressed) => {
```

Comment:
Fun fact: A raw BCF is actually invalid; the format is inherently BGZF-compressed. However, some libraries (e.g., HTSJDK) write raw BCF, and due to the output's prevalence, the VCF 4.4 spec now requires read support for it.

Reply:
Huh, I didn't know that - I'm too used to writing uncompressed BCF using `bcftools`.  Thanks for the tip!

Should we keep the support for uncompressed BCF in anticipation of VCF 4.4, or would you prefer that I remove it in the interrim?

Reply:
If you need uncompressed BCF, the spec recommends "files should be still compressed with BGZF but with compression 0" (§ 6.5 "BCF2 block gzip and indexing" (2022-08-22)).

Yes, keep it for interop and VCF 4.4 compliance.